pkg/controller/controller_instance.go

```diff
@@ -212,1168 +211,520 @@ func (c *controller) reconcileServiceInstanceKey(key string) error {
 	return c.reconcileServiceInstance(instance)
 }
 
-// reconcileServiceInstanceDelete is responsible for handling any instance whose
-// deletion timestamp is set.
-func (c *controller) reconcileServiceInstanceDelete(instance *v1beta1.ServiceInstance) error {
-	// nothing to do...
-	if instance.DeletionTimestamp == nil && !instance.Status.OrphanMitigationInProgress {
-		return nil
+// reconcileServiceInstance is the control-loop for reconciling Instances. An
+// error is returned to indicate that the instance has not been fully
+// processed and should be resubmitted at a later time.
+func (c *controller) reconcileServiceInstance(instance *v1beta1.ServiceInstance) error {
+	switch {
+	case instance.Status.AsyncOpInProgress:
+		return c.pollServiceInstance(instance)
+	case instance.ObjectMeta.DeletionTimestamp != nil || instance.Status.OrphanMitigationInProgress:
+		return c.reconcileServiceInstanceDelete(instance)
+	case instance.Status.ReconciledGeneration != 0:
+		return c.reconcileServiceInstanceUpdate(instance)
+	default:
+		return c.reconcileServiceInstanceAdd(instance)
 	}
+}
 
-	if finalizers := sets.NewString(instance.Finalizers...); !finalizers.Has(v1beta1.FinalizerServiceCatalog) {
+// reconcileServiceInstanceAdd is responsible for handling the provisioning
+// of new service instances.
+func (c *controller) reconcileServiceInstanceAdd(instance *v1beta1.ServiceInstance) error {
+	pcb := pretty.NewContextBuilder(pretty.ServiceInstance, instance.Namespace, instance.Name)
+
+	if isServiceInstanceFailed(instance) {
+		glog.V(4).Info(pcb.Message("Not processing event because status showed that it has failed"))
 		return nil
 	}
 
-	pcb := pretty.NewContextBuilder(pretty.ServiceInstance, instance.Namespace, instance.Name)
-
-	// If deprovisioning has failed, do not do anything more
-	if instance.Status.DeprovisionStatus == v1beta1.ServiceInstanceDeprovisionStatusFailed {
-		glog.V(4).Info(pcb.Message("Not processing deleting event because deprovisioning has failed"))
+	if instance.Status.ReconciledGeneration == instance.Generation {
+		glog.V(4).Info(pcb.Message("Not processing event because reconciled generation showed there is no work to do"))
 		return nil
 	}
 
-	glog.V(4).Info(pcb.Message("Processing deleting event"))
+	instance = instance.DeepCopy()
```

Comment:
Could the deep copy happen in reconcileServiceInstance?

Reply:
That's the eventual intent, but the tests rely on the various subfunctions having done the copying, so I had to duplicate it for now if I didn't want to touch the tests.